homeassistant/components/omnilogic/switch.py

```diff
@@ -0,0 +1,256 @@
+
+        await self.coordinator.api.set_relay_valve(
+            int(self._item_id[1]),
+            int(self._item_id[3]),
+            int(self._item_id[-1]),
+            1,
+        )
+
+    async def async_turn_off(self, **kwargs):
+        """Turn off the relay."""
+        self._state = False
+        self._last_action = time.time()
+        self.async_schedule_update_ha_state()
+
+        await self.coordinator.api.set_relay_valve(
+            int(self._item_id[1]),
+            int(self._item_id[3]),
+            int(self._item_id[-1]),
+            0,
+        )
+
+
+class OmniLogicPumpControl(OmniLogicSwitch):
+    """Define the OmniLogic Pump Switch Entity."""
+
+    def __init__(
+        self,
+        coordinator: OmniLogicUpdateCoordinator,
+        kind: str,
+        name: str,
+        icon: str,
+        item_id: tuple,
+        state_key: str,
+    ):
+        """Initialize entities."""
+        super().__init__(
+            coordinator=coordinator,
+            kind=kind,
+            name=name,
+            icon=icon,
+            item_id=item_id,
+            state_key=state_key,
+        )
+
+        self._max_speed = int(coordinator.data[item_id]["Max-Pump-Speed"])
+        self._min_speed = int(coordinator.data[item_id]["Min-Pump-Speed"])
+
+        if "Filter-Type" in coordinator.data[item_id]:
+            self._pump_type = PUMP_TYPES[coordinator.data[item_id]["Filter-Type"]]
+        else:
+            self._pump_type = PUMP_TYPES[coordinator.data[item_id]["Type"]]
+
+        self._last_speed = None
+
+    async def async_turn_on(self, **kwargs):
+        """Turn on the pump."""
+        self._state = True
+        self._last_action = time.time()
+        self.async_schedule_update_ha_state()
+
+        on_value = 100
+
+        if self._pump_type != "SINGLE" and self._last_speed:
+            on_value = self._last_speed
+
+        await self.coordinator.api.set_relay_valve(
+            int(self._item_id[1]),
+            int(self._item_id[3]),
+            int(self._item_id[-1]),
+            on_value,
+        )
+
+    async def async_turn_off(self, **kwargs):
+        """Turn off the pump."""
+        self._state = False
+        self._last_action = time.time()
+        self.async_schedule_update_ha_state()
+
+        if self._pump_type != "SINGLE":
+            if "filterSpeed" in self.coordinator.data[self._item_id]:
+                self._last_speed = self.coordinator.data[self._item_id]["filterSpeed"]
+            else:
+                self._last_speed = self.coordinator.data[self._item_id]["pumpSpeed"]
+
+        await self.coordinator.api.set_relay_valve(
+            int(self._item_id[1]),
+            int(self._item_id[3]),
+            int(self._item_id[-1]),
+            0,
+        )
+
+    async def async_set_speed(self, speed):
+        """Set the switch speed."""
+
+        if self._pump_type != "SINGLE":
+            if self._min_speed <= speed <= self._max_speed:
```

Comment:
Maybe this check should be in the library instead?

Reply:
Could be moved there if it is important. We chose to handle as part of the integration.

Reply:
OK, it's not critical.

Reply:
Looks like these two checks can be inverted and combined to form a guard clause that raises the exception. Then we can outdent below.